what4/src/What4/Protocol/SMTLib2.hs

```diff
@@ -884,6 +919,339 @@ parseBvArraySolverValue w v (SApp ["store", arr, idx, val]) = do
+  SApp [SAtom nm, SApp ["lambda", SApp params_sexp, body_sexp]] -> do
+    fn <- parseFn sym nm params_sexp body_sexp
+    return (nm, fn)
+  _ -> fail $ "unexpected sexp, expected lambda, found " ++ show sexp
+
+parseFn :: I.IsSymExprBuilder sym => sym -> Text -> [SExp] -> SExp -> IO (I.SomeSymFn sym)
+parseFn sym nm params_sexp body_sexp = do
+  (nms, vars) <- unzip <$> mapM (parseVar sym) params_sexp
+  case Ctx.fromList vars of
+    Some vars_assign -> do
+      let let_env = HashMap.fromList $ zip nms $ map (mapSome $ I.varExpr sym) vars
+      proc_res <- runProcessor (ProcessorEnv { procSym = sym, procLetEnv = let_env }) $ parseExpr sym body_sexp
+      Some body_expr <- either fail return proc_res
+      I.SomeSymFn <$> I.definedFn sym (I.safeSymbol $ Text.unpack nm) vars_assign body_expr I.NeverUnfold
+
+parseVar :: I.IsSymExprBuilder sym => sym -> SExp -> IO (Text, Some (I.BoundVar sym))
+parseVar sym sexp = case sexp of
+  SApp [SAtom nm, tp_sexp] -> do
+    Some tp <- parseType tp_sexp
+    var <- liftIO $ I.freshBoundVar sym (I.safeSymbol $ Text.unpack nm) tp
+    return (nm, Some var)
+  _ -> fail $ "unexpected variable " ++ show sexp
+
+parseType :: SExp -> IO (Some BaseTypeRepr)
+parseType sexp = case sexp of
+  "Bool" -> return $ Some BaseBoolRepr
+  "Int" -> return $ Some BaseIntegerRepr
+  "Real" -> return $ Some BaseRealRepr
+  SApp ["_", "BitVec", SAtom (Text.unpack -> m_str)]
+    | [(m_n, "")] <- readDec m_str
+    , Some m <- mkNatRepr m_n
+    , Just LeqProof <- testLeq (knownNat @1) m ->
+      return $ Some $ BaseBVRepr m
+  SApp ["_", "FloatingPoint", SAtom (Text.unpack -> eb_str), SAtom (Text.unpack -> sb_str)]
+    | [(eb_n, "")] <- readDec eb_str
+    , Some eb <- mkNatRepr eb_n
+    , Just LeqProof <- testLeq (knownNat @2) eb
+    , [(sb_n, "")] <- readDec sb_str
+    , Some sb <- mkNatRepr sb_n
+    , Just LeqProof <- testLeq (knownNat @2) sb ->
+      return $ Some $ BaseFloatRepr $ FloatingPointPrecisionRepr eb sb
+  SApp ["Array", idx_tp_sexp, val_tp_sexp] -> do
+    Some idx_tp <- parseType idx_tp_sexp
+    Some val_tp <- parseType val_tp_sexp
+    return $ Some $ BaseArrayRepr (Ctx.singleton idx_tp) val_tp
+  _ -> fail $ "unexpected type " ++ show sexp
+
+
+-- | Stores a NatRepr along with proof that its type parameter is a bitvector of
+-- that length. Used for easy pattern matching on the LHS of a binding in a
+-- do-expression to extract the proof.
+data BVProof tp where
+  BVProof :: forall n . (1 <= n) => NatRepr n -> BVProof (BaseBVType n)
+
+-- | Given an expression, monadically either returns proof that it is a
+-- bitvector or throws an error.
+getBVProof :: (I.IsExpr ex, MonadError String m) => ex tp -> m (BVProof tp)
+getBVProof expr = case I.exprType expr of
+  BaseBVRepr n -> return $ BVProof n
+  t -> throwError $ "expected BV, found " ++ show t
+
+-- | Operator type descriptions for parsing s-expression of
+-- the form @(operator operands ...)@.
+data Op sym where
```

Comment:
This looks incredibly similar to [`Op`](https://github.com/GaloisInc/what4/blob/17483418396afa8384de4089cdfaab9e5bbf912d/what4/src/What4/Serialize/Parser.hs#L271-L332) in `What4.Serialize.Parser`, to the point where I almost wonder if this code was copy-pasted. Why not just import `What4.Serialize.Parser` instead?

Reply:
this is copy-pasta code, with some SMT-LIB specific changes. I initially wanted to use `What4.Serialize.Parser`, but that's not exactly SMT-LIB, so the code there requires some work. I plan to merge this for now, to limit the scope of this change. todo https://github.com/GaloisInc/what4/issues/228

Reply:
OK. At a minimum, can you include a comment somewhere indicating:

1. That this is cargo-culted from `What4.Serialize.Parser`, and
2. A summary of the changes between it and the code in `What4.Serialize.Parser`